python/ray/data/_internal/execution/resource_manager.py

```diff
@@ -126,6 +144,50 @@ def _estimate_object_store_memory(self, op, state) -> int:
 
         return mem_op_internal + mem_op_outputs
 
+    def _estimate_object_store_memory_per_node(
+        self, op: "PhysicalOperator", state: "OpState"
+    ) -> Dict[str, int]:
+        usage_per_node = defaultdict(int)
+        # Don't count input refs towards dynamic memory usage, as they have been
+        # pre-created already outside this execution.
+        if isinstance(op, InputDataBuffer):
+            return usage_per_node
+
+        # Pending task outputs.
+        # TODO(mowen): We need something akin to op.metrics.obj_store_mem_pending_task_outputs
+        # here. Without the number of tasks running per node, this will be difficult to estimate.
+        # See: https://github.com/ray-project/ray/blob/b5bb747696300febb443c88a44938ac803103c25/python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py#L390
+
+        # Op's internal output buffers, iterate through internal outqueue
+        for (
+            node_id,
+            size,
+        ) in op.metrics._internal_outqueue.estimate_size_bytes_per_node().items():
+            usage_per_node[node_id] += size
+
+        # Op's external output buffer, iterate over outqueues in state.outqueue
+        for node_id, size in state.outqueue.memory_usage_per_node().items():
+            usage_per_node[node_id] += size
```

Comment:
@raulchen i'd rather not count this as an output queue, but rather as the input queue to make sure we're not counting these blocks twice.

WDYT?